google/cloud/storage/async_writer_test.cc

```diff
@@ -0,0 +1,156 @@
+  EXPECT_CALL(*mock, Write(WritePayloadContents(ElementsAre("bbb"))))
+      .WillOnce([] { return make_ready_future(Status{}); });
+  EXPECT_CALL(*mock, Finalize(WritePayloadContents(ElementsAre("ccc"))))
+      .WillOnce([] {
+        return make_ready_future(make_status_or(storage::ObjectMetadata()));
+      });
+
+  auto token = storage_internal::MakeAsyncToken(mock.get());
+  AsyncWriter writer(std::move(mock));
+  EXPECT_EQ(writer.UploadId(), "test-upload-id");
+  EXPECT_EQ(absl::get<std::int64_t>(writer.PersistedState()), 16384);
+  token = writer.Write(std::move(token), WritePayload{std::string("aaa")})
+              .get()
+              .value();
+  token = writer.Write(std::move(token), WritePayload{std::string("bbb")})
+              .get()
+              .value();
+  auto const actual =
+      writer.Finalize(std::move(token), WritePayload{std::string("ccc")}).get();
+  EXPECT_STATUS_OK(actual);
+}
+
+TEST(AsyncWriterTest, FinalizeEmpty) {
+  auto mock = std::make_unique<MockAsyncWriterConnection>();
+  ::testing::InSequence sequence;
+  EXPECT_CALL(*mock, Write(WritePayloadContents(ElementsAre("aaa"))))
+      .WillOnce([] { return make_ready_future(Status{}); });
+  EXPECT_CALL(*mock, Write(WritePayloadContents(ElementsAre("bbb"))))
+      .WillOnce([] { return make_ready_future(Status{}); });
+  EXPECT_CALL(*mock, Finalize(WritePayloadContents(IsEmpty()))).WillOnce([] {
+    return make_ready_future(make_status_or(storage::ObjectMetadata()));
+  });
+
+  auto token = storage_internal::MakeAsyncToken(mock.get());
+  AsyncWriter writer(std::move(mock));
+  token = writer.Write(std::move(token), WritePayload{std::string("aaa")})
+              .get()
+              .value();
+  token = writer.Write(std::move(token), WritePayload{std::string("bbb")})
+              .get()
+              .value();
+  auto const actual = writer.Finalize(std::move(token)).get();
+  EXPECT_STATUS_OK(actual);
+}
+
+TEST(AsyncWriterTest, ErrorDuringWrite) {
+  auto mock = std::make_unique<MockAsyncWriterConnection>();
+  EXPECT_CALL(*mock, Write).WillOnce([] {
+    return make_ready_future(PermanentError());
+  });
+
+  auto token = storage_internal::MakeAsyncToken(mock.get());
+  AsyncWriter writer(std::move(mock));
+  auto const actual = writer.Write(std::move(token), WritePayload{}).get();
+  EXPECT_THAT(actual, StatusIs(PermanentError().code()));
+}
+
+TEST(AsyncWriterTest, ErrorOnFinalize) {
+  auto mock = std::make_unique<MockAsyncWriterConnection>();
+  EXPECT_CALL(*mock, Finalize).WillOnce([] {
+    return make_ready_future(
+        StatusOr<storage::ObjectMetadata>(PermanentError()));
+  });
+
+  auto token = storage_internal::MakeAsyncToken(mock.get());
+  AsyncWriter writer(std::move(mock));
+  auto const actual = writer.Finalize(std::move(token), WritePayload{}).get();
+  EXPECT_THAT(actual, StatusIs(PermanentError().code()));
+}
+
+TEST(AsyncWriterTest, ErrorOnClosedStream) {
+  AsyncWriter writer;
```

Comment:
I am a little confused by the "Closed Stream" thing. Why do we even have a default constructor?

What should happen with `UploadId()` / `PersistedState()`? They do not handle errors `if (!impl)`.

Reply:
> I am a little confused by the "Closed Stream" thing. Why do we even have a default constructor?

You need it to support `std::tie()`, to put these into `std::pair` and it is the result of `std::move()`-from.

> 
> What should happen with `UploadId()` / `PersistedState()`? They do not handle errors `if (!impl)`.

Documented that as "undefined behavior".  The fact that I picked a specific behavior does not contradict the documentation.